[SPARK-39755][K8S] Improve `LocalDirsFeatureStep` to randomize local directories

pralabhkumar opened the pull request:
### What changes were proposed in this pull request?
Added randomization in spark local directories 


### Why are the changes needed?
In case of K8 , each executor gets same order  list of shuffle locations . For e.g if shuffle locations are s1 and s2 , then all the executor will get in the same order (s1,s2) .  Now based on the modulo , files are written in either S1 and S2. But for the same modulo file will always be written to the same location . For e.g if modulo is 0 file will be written to S1 and if it 1 then it will always be written to s2 in all the executors . This will lead to one disk being more utilized , then the other one . 

If we add randomization in the order of shuffle locations (as is done on yarn side) , executor will receive different order of shuffle locations  . Executor 1 can have (s1,s2) and executor 2 can have (s2, s1) . This will lead to fair utilization of disks . 

Randomization is already in place for yarn , this PR is to add the same in case resource manager is not yarn . 

Logs after changes 

Executor1 
```
22/07/13 03:45:06 INFO DiskBlockManager: Created local directory at /spark-local2/blockmgr-a08033da-0f9e-4c90-9c6a-b6ef600d545b
22/07/13 03:45:06 INFO DiskBlockManager: Created local directory at /spark-local1/blockmgr-1230c3e5-8c50-4e88-9dad-1954221e7228
```

Executor 2
```
22/07/13 03:45:08 INFO DiskBlockManager: Created local directory at /spark-local1/blockmgr-710a2944-3ce5-4a5b-9866-57708217f684
22/07/13 03:45:08 INFO DiskBlockManager: Created local directory at /spark-local2/blockmgr-2e2c79e2-9676-4c60-a94c-e4e9a9f6afd2

```

### Does this PR introduce _any_ user-facing change?
No


### How was this patch tested?
Added unit test cases also tested on internal K8s cluster.

pralabhkumar committed:
Randomization in Spark local directory for other resource managers

AmplabJenkins commented:
Can one of the admins verify this patch?

pralabhkumar committed:
fix in the failed unit test case

pralabhkumar commented:
@HyukjinKwon  Please review the PR . Thx

tgravescs commented:
> IMHO , this should be randomized , so that all the directories have equal changes of pushing the data as was done on yarn side

Was there an actual problem that occurred because these were not random?  

how was this tested?

pralabhkumar commented:
Please review the PR .

pralabhkumar commented:
> > IMHO , this should be randomized , so that all the directories have equal changes of pushing the data as was done on yarn side
> 
> Was there an actual problem that occurred because these were not random?

@tgravescs 
Yes we are seeing same problem(https://issues.apache.org/jira/browse/SPARK-24992) in our K8s cluster, where most of the time one disk get filled. 

> 
> how was this tested?


I had tested this via unit test cases . Also ran internally  on K8s cluster and seen executor logs . There have seen the disk get randomized on different executors.

dongjoon-hyun reviewed (commented):
Just a question. Why not applying for `SPARK_EXECUTOR_DIRS` code path together, @pralabhkumar ?

https://github.com/apache/spark/blob/66b1f79b72855af35351ff995492f2c13872dac5/core/src/main/scala/org/apache/spark/util/Utils.scala#L922-L923

And, could you elaborate your issue in the PR description more, please?

pralabhkumar committed:
Added randomization in spark_executor_dirs

tgravescs commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -937,7 +937,7 @@ private[spark] object Utils extends Logging {
       // configuration to point to a secure directory. So create a subdirectory with restricted
       // permissions under each listed directory.
-      conf.get("spark.local.dir", System.getProperty("java.io.tmpdir")).split(",")
+      randomizeInPlace(conf.get("spark.local.dir", System.getProperty("java.io.tmpdir")).split(","))
```
if we are calling it in all cases then we should just get the dirs and then call it at the end.

if val dirs = if (isRunningInYarnContainer(conf) {....}
// return:
randomizeInPlace(dirs)

cloud-fan replied:
+1

dongjoon-hyun replied:
+1

tgravescs commented:
ok to test

pralabhkumar commented:
I think build is failing because of unrelated changes (since test cases are passing in local) . Please let me know if I am missing something

pralabhkumar commented:
@dongjoon-hyun  I rebase and pushed the changes . However, I think  build is failing because of unrelated changes . 
Please help

HyukjinKwon commented:
Yeah, doesn't seem related. You can retrigger the failed tests for doubly sure rough

HyukjinKwon commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -920,9 +920,9 @@ private[spark] object Utils extends Logging {
       randomizeInPlace(getYarnLocalDirs(conf).split(","))
     } else if (conf.getenv("SPARK_EXECUTOR_DIRS") != null) {
-      conf.getenv("SPARK_EXECUTOR_DIRS").split(File.pathSeparator)
+      randomizeInPlace(conf.getenv("SPARK_EXECUTOR_DIRS").split(File.pathSeparator))
```
Hm, quick question though. `getConfiguredLocalDirs` gets invoked multiple times, and won't it return a different directory every time when it's called?

This `spark.local.dir` can be used to pass `spark.files` around, e.g., `SparkFiles.get`. Wonder if that would be affected.

If this is a K8S specific issue, I think it's best to localize the fix into K8S.

cloud-fan commented on core/src/test/scala/org/apache/spark/storage/LocalDirsSuite.scala:
```diff
@@ -79,10 +79,33 @@ class LocalDirsSuite extends SparkFunSuite with LocalRootDirsTest {
     assert(!f2.exists())
   }
+
+  test("SPARK-39755 : Test randomness in SPARK_LOCAL_DIRS, SPARK_EXECUTOR_DIRS") {
```
```suggestion
  test("SPARK-39755: Test randomness in SPARK_LOCAL_DIRS, SPARK_EXECUTOR_DIRS") {
```

pralabhkumar committed:
As suggested limited the scope only to K8s

HyukjinKwon commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -919,8 +925,13 @@ private[spark] object Utils extends Logging {
+    } else if (isRunningInK8sContainer(conf)) {
+      // Randomizing the shuffle location in case of K8s so that all disk get fair changes to
+      // get selected.
+      randomizeInPlace(conf.getenv("SPARK_LOCAL_DIRS").split(","))
```
I am still not sure if this really safe to return the randomized directory whenever this method is invoked. This configuration (spark.local.dir) and environment variables are meant to be immutable but this is changing it here for every call during runtime.

Since this relates to the code path in which we should be relatively more strict, I would suggest to thoroughly investigate all the related paths TBH since we're breaking the assumption.

mridulm replied:
QQ: I see both `SPARK_LOCAL_DIRS` and `SPARK_EXECUTOR_DIRS` in use within k8s resource manager - do we not want to do this for `SPARK_EXECUTOR_DIRS` @dongjoon-hyun ?

dongjoon-hyun replied:
> QQ: I see both `SPARK_LOCAL_DIRS` and `SPARK_EXECUTOR_DIRS` in use within k8s resource manager - do we not want to do this for `SPARK_EXECUTOR_DIRS` @dongjoon-hyun ?

We want it, @mridulm . That was my first comment here.
- https://github.com/apache/spark/pull/37203#pullrequestreview-1045944162

pralabhkumar committed:
Handling case when spark_local directories are not set

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -919,8 +925,13 @@ private[spark] object Utils extends Logging {
+      randomizeInPlace(conf.getenv("SPARK_LOCAL_DIRS").split(","))
+    }
+    else if (conf.getenv("SPARK_EXECUTOR_DIRS") != null) {
+     conf.getenv("SPARK_EXECUTOR_DIRS").split(File.pathSeparator)
```
Please recover indentation change here.

pralabhkumar replied:
will do

dongjoon-hyun commented on core/src/test/scala/org/apache/spark/storage/LocalDirsSuite.scala:
```diff
@@ -85,4 +85,24 @@ class LocalDirsSuite extends SparkFunSuite with LocalRootDirsTest {
+    assert(index_of_path1(0))
+    assert(index_of_path1(1))
+  }
+
```
nit. redundant empty line.

pralabhkumar replied:
will do

dongjoon-hyun commented:
To @pralabhkumar . The following two comments seems to be not applied properly at the final commit. Is it applied and reverted?
- https://github.com/apache/spark/pull/37203#pullrequestreview-1045944162 (mine)
- https://github.com/apache/spark/pull/37203#discussion_r927667939 (@tgravescs 's)

pralabhkumar commented:
> To @pralabhkumar . The following two comments seems to be not applied properly at the final commit. Is it applied and reverted?
> 
> * [[SPARK-39755][CORE] Randomization in Spark local directory for K8 resource managers #37203 (review)](https://github.com/apache/spark/pull/37203#pullrequestreview-1045944162) (mine)
> * [[SPARK-39755][CORE] Randomization in Spark local directory for K8 resource managers #37203 (comment)](https://github.com/apache/spark/pull/37203#discussion_r927667939) (@tgravescs 's)

@dongjoon-hyun 
Yes the first comment was applied , however @HyukjinKwon suggested to do the changes only w.r.t to K8s , therefore I had reverted it , because of which I didn't apply the second one .

pralabhkumar committed:
Indentation changes

pralabhkumar committed:
rectified Indentation

mridulm commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -880,6 +880,12 @@ private[spark] object Utils extends Logging {
+  private[spark] def isRunningInK8sContainer(conf: SparkConf): Boolean = {
+    // Master must start with K8s in case of kubernetes
+    conf.contains("spark.master") && conf.get("spark.master") != null &&
+      conf.get("spark.master").startsWith("k8s://")
```
nit: `conf.getOption("spark.master").exists(_.startsWith("k8s://"))` ?

pralabhkumar committed:
Addressed review comments

pralabhkumar commented:
@mridulm @cloud-fan @tgravescs @HyukjinKwon @dongjoon-hyun 

Have addressed all the review comments. Please guide me to take this further .

pralabhkumar commented:
@dongjoon-hyun can u please review.

pralabhkumar commented:
Gentle ping to review the PR

pralabhkumar commented:
@HyukjinKwon @cloud-fan @dongjoon-hyun Please find some time to review the PR .

pralabhkumar commented:
@HyukjinKwon @dongjoon-hyun Ping to please review the PR and take it forward

dongjoon-hyun reviewed (commented):
Sorry for being late, @pralabhkumar . I'll resume the review in one or two days for this PR.

dongjoon-hyun reviewed (changes requested):
I realized that this is still partial and might not be the best place to have K8s specific code, @pralabhkumar . For example, K8s `LocalDirsFeatureStep` uses `SPARK_LOCAL_DIRS` or `spark.local.dir` environment variable directly still instead of using `Utils.getConfiguredLocalDirs`.

https://github.com/apache/spark/blob/afd7098c7fb6c95aece39acc32cdad764c984cd2/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala#L43-L51

Can we have a test coverage in K8s unit tests or integration tests?

pralabhkumar commented:
@dongjoon-hyun . Thx for reviewing it .  Let me rework on the PR as suggested.

pralabhkumar committed:
Implemented suggested changes

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala:
```diff
@@ -24,6 +24,7 @@ import io.fabric8.kubernetes.api.model._
 
 import org.apache.spark.deploy.k8s.{KubernetesConf, SparkPod}
 import org.apache.spark.deploy.k8s.Config._
+import org.apache.spark.util.Utils._
```
Please update like the following.
```scala
- import org.apache.spark.util.Utils._
+ import org.apache.spark.util.Utils.randomizeInPlace
```

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala:
```diff
@@ -35,9 +36,10 @@ private[spark] class LocalDirsFeatureStep(
     var localDirs = pod.container.getVolumeMounts.asScala
       .filter(_.getName.startsWith("spark-local-dir-"))
-      .map(_.getMountPath)
+      .map(_.getMountPath).toArray
```
BTW, `Utils` has the following too.

```
/**
   * Shuffle the elements of a collection into a random order, returning the
   * result in a new collection. Unlike scala.util.Random.shuffle, this method
   * uses a local random number generator, avoiding inter-thread contention.
   */
  def randomize[T: ClassTag](seq: TraversableOnce[T]): Seq[T] = {
    randomizeInPlace(seq.toArray)
  }
```

dongjoon-hyun reviewed (changes requested):
BTW, we need a new test coverage for this PR, @pralabhkumar .

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala:
```diff
@@ -35,9 +36,10 @@ private[spark] class LocalDirsFeatureStep(
+      .map(_.getMountPath).toArray
     var localDirVolumes : Seq[Volume] = Seq()
     var localDirVolumeMounts : Seq[VolumeMount] = Seq()
+    var isShuffleLocationRandomized : Boolean = false
```
Why do we have this? You can randomize `localDirs` at line 37.

pralabhkumar committed:
Addressed review comments and added unit test case

dongjoon-hyun commented on core/src/test/scala/org/apache/spark/storage/LocalDirsSuite.scala:
```diff
@@ -85,4 +85,5 @@ class LocalDirsSuite extends SparkFunSuite with LocalRootDirsTest {
     assert(!f1.exists())
     assert(!f2.exists())
   }
+
```
Please revert this, @pralabhkumar .

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala:
```diff
@@ -33,12 +34,11 @@ private[spark] class LocalDirsFeatureStep(
-      .map(_.getMountPath)
-    var localDirVolumes : Seq[Volume] = Seq()
-    var localDirVolumeMounts : Seq[VolumeMount] = Seq()
-
```
Please revert this removal.

pralabhkumar replied:
These line are not removed its there on line 40 and 41 .

dongjoon-hyun replied:
What I pointed was the removal of that empty line after `var localDirVolumeMounts`.

<img width="1202" alt="Screen Shot 2022-08-20 at 12 45 14 PM" 
src="https://user-images.githubusercontent.com/9700541/185763765-0be7e2df-b13d-4ecf-83b2-f4af6079e408.png">

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala:
```diff
@@ -68,9 +69,8 @@ private[spark] class LocalDirsFeatureStep(
-          }
+        }
     }
-
```
Please revert this removal.

pralabhkumar replied:
done

dongjoon-hyun commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStepSuite.scala:
```diff
@@ -52,41 +52,58 @@ class LocalDirsFeatureStepSuite extends SparkFunSuite {
   }
 
   test("Use configured local dirs split on comma if provided.") {
```
Please avoid changing the existing test coverage. Instead, we recommend to add a new test coverage as a new test case with test name prefix, `SPARK-39755: `. Technically, this test case doesn't fail with your PR, does it?

pralabhkumar replied:
@dongjoon-hyun  After randomization , this test case will fail , since this test case is always expecting 
```
assert(configuredPod.container.getVolumeMounts.get(0) ===
      new VolumeMountBuilder()
        .withName(s"spark-local-dir-1")
        .withMountPath("/var/data/my-local-dir-1")
        .build())
```

which is not true always since , locations are randomized.  Also  test case is expecting env variable as always 
/var/data/my-local-dir-1,/var/data/my-local-dir-2 which is not true after randomization. 

```
assert(configuredPod.container.getEnv.get(0) ===
      new EnvVarBuilder()
        .withName("SPARK_LOCAL_DIRS")
        .withValue("/var/data/my-local-dir-1,/var/data/my-local-dir-2")
        .build())
```

Therefore have changed the test case to take care of randomization.

dongjoon-hyun replied:
In that case, simply loose the index condition to check `.get(0)` and `get(1)` instead of refactoring like this.

pralabhkumar committed:
Reverted irrelvant changes

pralabhkumar committed:
simplified unit test case

pralabhkumar committed:
Simplified unit test case for randomization

pralabhkumar committed:
Fixing the error in unit test case

pralabhkumar commented:
@dongjoon-hyun  fixed all comments and build is also passing . please review

dongjoon-hyun commented:
Sure, I'll recisit this Today.

dongjoon-hyun reviewed (commented):
1. Did you test this patch to your cluster, @pralabhkumar ? I'm wondering if this fixes your problems or not.
2. For `SPARK_EXECUTOR_DIRS`, we can handle it later in a new JIRA issue separately.

pralabhkumar commented:
@dongjoon-hyun . Thx for reviewing. 
1. Yes we tested in our cluster . It helps to randomly pic shuffle locations , which helps in equal  data distribution for disks .
2. I'll create a seperate Jira for the SPARK_EXECUTOR_DIRS .

dongjoon-hyun reviewed (approved):
Thank you, @pralabhkumar . That's enough for this PR.
+1, LGTM. Merged to master for Apache Spark 3.4.0.

dongjoon-hyun commented:
Thank you, @tgravescs , @mridulm , @cloud-fan , @HyukjinKwon , too.
This is a pure K8s module PR now and landed to Apache Spark 3.4.0.

pralabhkumar commented:
Thx for all the help provided during review @dongjoon-hyun
